Use text properties instead of overlays for ANSI coloring

rfkm opened the pull request:
### TL;DR

This PR improves the performance of REPL buffers by using text properties instead of overlays for ANSI coloring.
Also, this fixes #1452 while removing the hacky workaround introduced by #1455.

---

I recently sent some PRs related to overlays in the REPL buffer,
but I've finally realized we should avoid using overlays for ANSI coloring in the first place.
It's simply because overlays are pretty slow. I improved the performance a bit in #1488, but it's not radical.

So, in this PR, I remove all code related to overlays and use text properties instead.


I measured the performance by calling `cider-repl--emit-interactive-output` repeatedly.
The complete benchmark code is [here](https://gist.github.com/rfkm/c28f0e35965bcebe3e37), and the version of my Emacs is `GNU Emacs 24.4.1 (x86_64-apple-darwin14.0.0, NS apple-appkit-1343.14)`.

![active](https://cloud.githubusercontent.com/assets/857162/12069428/6520b1d2-b072-11e5-914b-536ec23da79a.png)

As this figure shows, the overlay version (current master) is getting slower and slower as the number of lines in the REPL buffer increases.

![total](https://cloud.githubusercontent.com/assets/857162/12069432/fe855832-b072-11e5-9fe3-d8adf4be0b94.png)

Total time the overlay version spent on inserting 1000 lines is about 5 seconds. As for the text property version, it's about 0.6 seconds.

If no window displays the REPL buffer, it is much faster in the both cases:

![hidden](https://cloud.githubusercontent.com/assets/857162/12069434/0dc1b3ae-b073-11e5-9cd2-4c0e77a53796.png)

![total_hidden](https://cloud.githubusercontent.com/assets/857162/12069435/138ed000-b073-11e5-8496-f8dfc822467a.png)

---

I keep my changes in #1488 as is because they are harmless, but I can remove them if it is necessary.

I also wonder whether I should remove my old changelog entries related to overlays.
Since this PR removes old PRs' changes, it looks a bit strange that these changelog entries appears in the same release.


P.S.

While I had not intended that, the PR #1455 seems to have improved the performance significantly ;)

![active_old](https://cloud.githubusercontent.com/assets/857162/12069438/64297ba0-b073-11e5-87b3-e07a4425503d.png)
(This is the worst case, that is, all outputs have a trailing overlay.)

Malabarba commented:
Makes sense to me. Overlays are indeed very slow in large numbers. 
Thanks for the very thorough analysis. :-)

Do we have tests for this ansi coloring feature?

bbatsov commented:
> I also wonder whether I should remove my old changelog entries related to overlays.
> Since this PR removes old PRs' changes, it looks a bit strange that these changelog entries appears in the same release.

Yeah, you should remove them.

Overall - impressive work. I was amazed by your comprehensive analysis of the problem!
Performance optimization work is critical, but we've always ignored it to some extent so far. Maybe you can help us with tickets like:

* https://github.com/clojure-emacs/cider/issues/228
* https://github.com/clojure-emacs/cider/issues/1115

> Do we have tests for this ansi coloring feature?

Pretty sure we don't. A few tests would be nice to have.

Malabarba commented:
Also, I forgot to mention, but the code is good to merge for me (though some tests would be nice).

bbatsov commented:
Yeah, it's good to merge for sure - I just want us to update the changelog accordingly as well.

rfkm commented:
Sorry for the delay.
I've just pushed some changes.
Since this PR also fixes #1452, I kept the changelog entry about #1452 though I removed the mention of overlays.
Instead, I've added summary of this PR on top of this thread.
Also, I added a test for this.

> Performance optimization work is critical, but we've always ignored it to some extent so far. Maybe you can help us with tickets like:

I've already noticed `cider-repl--show-maximum-output` could be slow under certain conditions, but I hope one of the causes has been already eliminated.
One of the causes I noticed is `(recenter -1)` that `cider-repl--show-maximum-output` calls could be pretty slow if there are many overlays.
Also, I've noticed eldoc could be very slow in a large REPL buffer. As for this, I might send a PR in the near future.
Anyway, I'll try to ivenstigate their issues :)

bbatsov commented:
Great! Thanks a lot! :-)

Malabarba commented:
Thanks indeed!
